### PR TITLE
proposal

### PR DESCRIPTION
**Pull request by F4ever**

### CODE REVIEW

File: PROPOSALCOPY.MD [hunk 1]
```diff
@@ -0,0 +1,42 @@
+## Proposal to use a bot to deposit keys and ether to contract.
+
+As you know, to keep our stacking protocol working we should deposit keys and ether to deposit contract.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Many wouldn't know technical details, especially as deposits were triggered by the dev team's hands only.

Comment (kadmil) [reply PROPOSALCOPY.MD]: Keys aren't deposited, the ether is

Comment (kadmil) [reply PROPOSALCOPY.MD]: Users' ether submitted to Lido smart contract is deposited to the Beacon chain staking via `depositBufferedEther` call on Lido smart contract. This operation requires significant amount of gas, making manual operations not as cost-efficient as possible. We propose automating the deposits with bot, implementing smart gas strategy suitable for task's constraints.

Comment (kadmil) [reply PROPOSALCOPY.MD]: Along with this, we propose refunding dev team for gas spent on deposits so far (X ether)

Comment (kadmil) [reply PROPOSALCOPY.MD]: (need to mention bot funding by DAO as well)

File: PROPOSALCOPY.MD [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+For now, we don't have any automatic solution.
+
+We propose:
+1. Run and fund bot that will deposit buffered ether to contract.
+2. Make refund to the development team that deposited buffered ether to contract before.
+
+## Problem
+
+Ether is placed in buffer, and it must be deposited by calling smart contract. Until now, it was done manually.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: The problem isn't in the buffer, but in that: 1) it requires sc interaction to deposit; 2) protocol apr is earned only on deposited ether — having big buffer hurts apr; 3) the operation is gas-consuming, making it desirable to choose some form of optimal moment / gas price to a) not overspend; b) don't have too much ether in the buffer for too long, hurting the protocol's apr

File: PROPOSALCOPY.MD [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+Manual action requires attention, and we can save money if we deposit while the cost of gas is low.
+
+## Strategy
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Maybe the `Proposed solution` title would fit better?

File: PROPOSALCOPY.MD [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+The main idea is to control gas fee and create transaction when the gas fee is quite low.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: It's better not to use relative terms without numbers.

Comment (kadmil) [reply PROPOSALCOPY.MD]: Another thing, if that's a section to outline suggested solution, we may write out what the solution is — not the gas strategy per se, but the bot utilizing it + the infra required to get the relevant gas price data

File: PROPOSALCOPY.MD [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+One of strategy is to count the 20 percentile for the previous couple of days.  
+We calculate two values:
+- 20 percentile for 1 day
+- 20 percentile for 4 days
+And the minimum of them becomes the recommended fee.
+
+If current fee is lower than recommended fee we submit transactions.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: ... and if there's ether in the protocol buffer

File: PROPOSALCOPY.MD [hunk 6]
```diff
@@ -0,0 +1,42 @@
+## Proposal to use a bot to deposit keys and ether to contract.
+
+As you know, to keep our stacking protocol working we should deposit keys and ether to deposit contract.
+For now, we don't have any automatic solution.
+
+We propose:
+1. Run and fund bot that will deposit buffered ether to contract.
+2. Make refund to the development team that deposited buffered ether to contract before.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Need a number there, how much ether has been spent?

File: PROPOSALCOPY.MD [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+We calculate two values:
+- 20 percentile for 1 day
+- 20 percentile for 4 days
+And the minimum of them becomes the recommended fee.
+
+If current fee is lower than recommended fee we submit transactions.
+
+## Strategy details
+
+I wrote a small script that looks at historical data and plots charts with recommended price, percentile for 1 day and percentile for 4 days. And got this:
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Better just present result, like `Chart with 1-day & 4-days 20th percentiles`

Comment (kadmil) [reply PROPOSALCOPY.MD]: And update the chart before submitting

File: PROPOSALCOPY.MD [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+If current fee is lower than recommended fee we submit transactions.
+
+## Strategy details
+
+I wrote a small script that looks at historical data and plots charts with recommended price, percentile for 1 day and percentile for 4 days. And got this:  
+
+<img src="https://github.com/F4ever/gas-strategy/blob/master/plot_example.png" alt="Gas fee chart">
+  
+You can play with sandbox and change percentile and/or period used to calculate percentile:
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Again, better leave a link and tell what it is; "Script used to build a plot: ..."

File: PROPOSALCOPY.MD [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+I wrote a small script that looks at historical data and plots charts with recommended price, percentile for 1 day and percentile for 4 days. And got this:  
+
+<img src="https://github.com/F4ever/gas-strategy/blob/master/plot_example.png" alt="Gas fee chart">
+  
+You can play with sandbox and change percentile and/or period used to calculate percentile:  
+[Gas strategy proposal](https://github.com/F4ever/gas-strategy)
+
+## Implementation
+
+Here is the source for the depositor bot: [Depositor source code](https://github.com/lidofinance/depositor-bot)
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: The proposed bot source code: ...

File: PROPOSALCOPY.MD [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+## Compensation
+
+Ether calculations that the development team spent to deposit ether: [https://dune.xyz/queries/162545](https://dune.xyz/queries/162545)
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Need a number here as well

File: PROPOSALCOPY.MD [hunk 11]
```diff
@@ -0,0 +1,48 @@
+## Proposal to use a bot to deposit keys and ether to contract.
+
+Users' ether submitted to Lido smart contract is deposited to the Beacon chain staking via depositBufferedEther call on Lido smart contract. 
+This operation requires significant amount of gas, making manual operations not as cost-efficient as possible.  
+
+We propose to run and fund bot that will automate the deposits, 
+implementing smart gas strategy suitable for task's constraints.
+Along with this, we propose refunding dev team for gas spent on deposits so far (86.16 ETH)
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: May set a full amount with WEIs; note that the number has changed already

File: PROPOSALCOPY.MD [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+User's ether submitted to Lido smart contract is deposited to the Beacon chain staking via depositBufferedEther call on Lido smart contract. 
+This operation requires significant amount of gas, making manual operations not as cost-efficient as possible.  
+
+We propose to run and fund bot that will automate the deposits, implementing smart gas strategy suitable for task's constraints.
+Along with this, we propose refunding dev team for gas spent on deposits so far (86.16 ETH)
+
+## Problem
+
+Protocol's APR is earned only on deposited ether - having big buffer hurts APR. 
+It requires smart contract interaction to deposit ether.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Ether is deposited through the smart contract call (`depostiBufferedEther` on Lido smart contract `0xdcd90c7f6324cfa40d7169ef80b12031770b4325`)

File: PROPOSALCOPY.MD [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+This operation requires significant amount of gas, making manual operations not as cost-efficient as possible.  
+
+We propose to run and fund bot that will automate the deposits, implementing smart gas strategy suitable for task's constraints.
+Along with this, we propose refunding dev team for gas spent on deposits so far (86.16 ETH)
+
+## Problem
+
+Protocol's APR is earned only on deposited ether - having big buffer hurts APR. 
+It requires smart contract interaction to deposit ether. 
+Deposit ether is gas-consuming, making it desirable to choose some form of optimal moment / gas price to:
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: The ether deposit is a gas-intensive operation, so it should be done balancing gas expenses and the time ether sits in protocol buffer.

(may get rid of the list below)

File: PROPOSALCOPY.MD [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+1. not overspend;
+2. don't have too much ether in the buffer for too long, hurting the protocol's apr;
+
+## Proposed solution
+
+Write bot that will track gas fee history and deposit buffered ether if conditions are met (gas fee, buffered ether, unused keys).
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Implement and deploy a bot which would deposit buffered ether using the specified gas strategy.

File: PROPOSALCOPY.MD [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+#### Gas strategy:
+
+One of strategy we propose is to count the 20 percentiles for the previous couples of days and use one of them as recommended fee.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: The proposed strategy is to send deposit transactions once the current gas price is below both 1-day 20-th percentile & 4-days 20-th percentile.

Comment (kadmil) [reply PROPOSALCOPY.MD]: The rationale behind this choice is required

File: PROPOSALCOPY.MD [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+- 20 percentile for 1 day
+- 20 percentile for 4 days
+
+And the minimum of them becomes the recommended fee.
+
+As soon as current fee is lower than recommended fee we allow submitting deposit transaction.
+
+## Strategy details
+
+Chart with 1-day & 4-days 20th percentiles:
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Need an update before publishing

File: PROPOSALCOPY.MD [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+Calculate two values:
+- 20 percentile for 1 day
+- 20 percentile for 4 days
+
+And the minimum of them becomes the recommended fee.
+
+As soon as current fee is lower than recommended fee we allow submitting deposit transaction.
+
+## Strategy details
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Not sure it's "details" — it's more like a display/justification for the rationale behind the choises

File: PROPOSALCOPY.MD [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+Script used to build a plot:
+[Gas strategy proposal](https://github.com/F4ever/gas-strategy)
+
+## Implementation
+
+The proposed bot source code: [Depositor source code](https://github.com/lidofinance/depositor-bot)
+
+## Compensation
+
+Total ether spent by dev team on gas: 86.16 ETH.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Need a preamble of "dev team has been running the operation manually and has spent X ether while doing so"...

File: PROPOSALCOPY.MD [hunk 19]
```diff
@@ -0,0 +1,41 @@
+## Proposal to use a bot to deposit keys and ether to contract.
+
+User's ether submitted to Lido smart contract is deposited to the Beacon chain staking via depositBufferedEther call on Lido smart contract.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Let's get rid of the first "smart contract" here

File: PROPOSALCOPY.MD [hunk 20]
```diff
@@ -0,0 +1,41 @@
+## Proposal to use a bot to deposit keys and ether to contract.
+
+User's ether submitted to Lido smart contract is deposited to the Beacon chain staking via depositBufferedEther call on Lido smart contract. 
+This operation requires significant amount of gas, making manual operations not as cost-efficient as possible.  
+
+We propose to run and fund bot that will automate the deposits, implementing smart gas strategy suitable for task's constraints.
+Along with this, we propose refunding dev team for gas spent on deposits so far (87,814,430,128,333,960,000 WEI; note that the number may have changed)
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: 87.81443012833396 ETH

Comment (kadmil) [reply PROPOSALCOPY.MD]: don't add "note that the number may have changed — it's internal note

Comment (kadmil) [reply PROPOSALCOPY.MD]: don't add "note that the number may have changed — it's internal note

File: PROPOSALCOPY.MD [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+#### Gas strategy:
+
+The proposed strategy is to send deposit transactions once the current gas price is below both 1-day 20-th percentile & 4-days 20-th percentile.
+
+## Strategy justification
+
+Chart with 1-day & 4-days 20th percentiles (1 week long):  
+<img src="https://github.com/F4ever/gas-strategy/blob/master/plot_example.png" alt="Gas fee chart">
+  
+As you can see 4-day percentile prevents the recommended fee to rise sharply.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Let's remove "as you can see"

File: PROPOSALCOPY.MD [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+In same time 1-day percentile perfectly counts recommended fee if gas fee goes down.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: not "counts", but allows to deposit cheaper if the gas price goes down

File: PROPOSALCOPY.MD [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+Script used to build a plot:
+[Gas strategy proposal](https://github.com/F4ever/gas-strategy)
+
+## Implementation
+
+The proposed bot source code: [Depositor source code](https://github.com/lidofinance/depositor-bot)
+
+## Compensation
+
+Dev team has been running the operation manually and has spent 87.81 ether.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: The sum to be the same as above

File: PROPOSALCOPY.MD [hunk 24]
```diff
@@ -0,0 +1,41 @@
+## Proposal to use a bot to deposit keys and ether to contract.
```
Comment (kadmil) [inline PROPOSALCOPY.MD]: Deploy & fund the bot for depositing buffered ether

**Comment by vshvsh** [comment]
1. I think it would make sense to split up the "already spent" eth from funding future spends, and have them in different proposals
2. Future spends should come with outline of the gas strategy we will use in more details, preferably as an open source depositor bot (link to the strategy itself, the picture you made would also look good); also it makes sense to make it open discussion - maybe we (Lido?) want to incentivize other people to run the bot rather than running it by devs?

The proposal is awfully short on details (who was running the bot before? how 100+ eth were spent like this? who is being refunded? how do we make sure right people are reefunded?), justifications and overall looks drafty.

File: main.py [hunk 25]
```diff
@@ -10,6 +10,9 @@
 
 
 INFURA_PROJECT_ID = os.getenv('INFURA_PROJECT_ID')
+if INFURA_PROJECT_ID is None:
+    raise Exception('Provide INFURA_PROJECT_ID in environment')
+
```
Comment (kadmil) [inline main.py:15]: Needs to be a separate pull

File: PROPOSALCOPY.MD [hunk 26]
```diff
@@ -0,0 +1,36 @@
+## Proposal to deploy & fund the bot for depositing buffered ether and refund ether spent on gas fee
+
+User's ether is deposited to the Beacon chain staking via depositBufferedEther call on "DepositSecurityModule" smart contract.
```
Comment (kadmil) [inline PROPOSALCOPY.MD:3]: The motivation is required; another thing to mention here is: we _need_ a bot for vulnerability mitigation, at least for the current implementation (one can't read from Deposit Security Council Kafka by hand)

File: PROPOSALCOPY.MD [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+This operation requires significant amount of gas and is performed by the depositor bot.
+
+We propose to fund the depositor bot and refund dev team for gas spent on deposits so far (124.72235422715326 eth)
```
Comment (kadmil) [inline PROPOSALCOPY.MD:6]: Let's separate those two in different sentences. please =)

File: PROPOSALCOPY.MD [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+
+## Problem
+
+Protocol's APR is earned only on deposited ether - having big buffer hurts APR.
+
+Ether is deposited through the smart contract call (depostiBufferedEther on DepositSecurityModule smart contract 0xDb149235B6F40dC08810AA69869783Be101790e7),
```
Comment (kadmil) [inline PROPOSALCOPY.MD:12]: typo `depostiBufferedEther`

File: PROPOSALCOPY.MD [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+
+We must periodically replenish the bot's wallet to continue doing deposits.
+
+## Proposed solution
+
+Fund the bot that would deposit buffered ether.
+
+#### Gas costs
+
+To reduce gas costs we use special gas strategy.
```
Comment (kadmil) [inline PROPOSALCOPY.MD:25]: Need an estimate: how much ETH is required to deposit 1k ETH buffered? 5k? What's the ballpark for monthly expenses?

File: PROPOSALCOPY.MD [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+The strategy is to send deposit transactions once the current gas price is below 2-day 30-th percentile.
```
Comment (kadmil) [inline PROPOSALCOPY.MD:26]: Why this number? Can we provide motivation / calc behind it?

**Review by kadmil — COMMENTED** [review]
Sent comments; the gist of it is, we want to provide motivation & calc behind the proposal & numbers outlined there. May not have those rn, but would want to figure them out before the posting

**Pull request closed without merging**